// rc-disco/dist/Layout.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('./styles');

var _styles2 = _interopRequireDefault(_styles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Layout = function Layout(_ref) {
  var opacity = _ref.opacity,
      duration = _ref.duration,
      deg = _ref.deg,
      gradient = _ref.gradient,
      onTransitionEnd = _ref.onTransitionEnd;
  return _react2.default.createElement('div', {
    onTransitionEnd: onTransitionEnd,
    style: _extends({}, _styles2.default.layout, {
      opacity: opacity,
      transitionDuration: duration + 'ms',
      background: 'linear-gradient(' + deg + 'deg, ' + gradient.colors + ')'
    })
  });
};

Layout.propTypes = {
  onTransitionEnd: _propTypes2.default.func,
  opacity: _propTypes2.default.number,
  duration: _propTypes2.default.number,
  deg: _propTypes2.default.number,
  gradient: _propTypes2.default.object.isRequired
};

exports.default = Layout;
